core/services/bulletprooftxmanager/nonce_syncer.go

```diff
@@ -0,0 +1,406 @@
+	"github.com/smartcontractkit/chainlink/core/utils"
+	"go.uber.org/multierr"
+	"gorm.io/gorm"
+)
+
+type (
+	// NonceSyncer manages the delicate task of syncing the local nonce with the
+	// chain nonce in case of divergence.
+	//
+	// On startup, we check each key for the nonce value on chain and compare
+	// it to our local value.
+	//
+	// Usually the on-chain nonce will be the same as (or lower than) the
+	// next_nonce in the DB, in which case we do nothing.
+	//
+	// If we are restoring from a backup however, or another wallet has used the
+	// account, the chain nonce might be higher than our local one. In this
+	// scenario, we must fastforward the local nonce to match the chain nonce.
+	//
+	// The problem with doing this is that now Chainlink does not have any
+	// ownership or control over potentially pending transactions with nonces
+	// between our local highest nonce and the chain nonce. If one of those
+	// transactions is pushed out of the mempool or re-org'd out of the chain,
+	// we run the risk of being stuck with a gap in the nonce sequence that
+	// will never be filled.
+	//
+	// The solution is to query the chain for our own transactions and take
+	// ownership of them by writing them to the database and letting the
+	// EthConfirmer handle them as it would any other transaction.
+	//
+	// This is not quite as straightforward as one might expect. We cannot
+	// query transactions from our account to infinite depth (geth does not
+	// support this). The best we can do is to query for all transactions sent
+	// within the past ETH_FINALITY_DEPTH blocks and find the ones sent by our
+	// address(es).
+	//
+	// This gives us re-org protection up to ETH_FINALITY_DEPTH deep in the
+	// worst case, which is in line with our other guarantees.
+	NonceSyncer struct {
+		store     *store.Store
+		config    orm.ConfigReader
+		ethClient eth.Client
+	}
+	// NSinserttx represnets an EthTx and Attempt to be inserted together
+	NSinserttx struct {
+		Etx     models.EthTx
+		Attempt models.EthTxAttempt
+	}
+)
+
+// NewNonceSyncer returns a new syncer
+func NewNonceSyncer(store *store.Store, config orm.ConfigReader, ethClient eth.Client) *NonceSyncer {
+	return &NonceSyncer{
+		store,
+		config,
+		ethClient,
+	}
+}
+
+// SyncAll syncs nonces for all keys in parallel
+//
+// This should only be called once, before the EthBroadcaster has started.
+// Calling it later is not safe and could lead to races.
+func (s NonceSyncer) SyncAll(ctx context.Context) (merr error) {
+	keys, err := s.store.SendKeys()
+	if err != nil {
+		return errors.Wrap(err, "NonceSyncer#fastForwardNoncesIfNecessary failed to get keys")
+	}
+
+	var wg sync.WaitGroup
+	var errMu sync.Mutex
+
+	wg.Add(len(keys))
+	for _, key := range keys {
+		go func(k models.Key) {
+			defer wg.Done()
+			if err := s.fastForwardNonceIfNecessary(ctx, k.Address.Address()); err != nil {
+				errMu.Lock()
+				defer errMu.Unlock()
+				merr = multierr.Combine(merr, err)
+			}
+		}(key)
+	}
```

Comment:
just a notice - could be a nice util function `ParallelProcess(slice []interface{}, process func(item interface{}) error`  uh if only we had generics 😅

Reply:
alternatively, i guess could use channels for passing the errors back

Reply:
Don't you like the mutexes?

Reply:
in Scala there were usually ways to avoid them, in Go i don't have an opinion yet ;)